Initial configuration that runs full closed loop with cwfs.

jbkalmbach opened the pull request.

jbkalmbach committed:
Initial configuration that runs full closed loop with cwfs.

jbkalmbach committed:
Move to only 4 fields.

jbkalmbach committed:
Latest changes for cwfs loop.

jbkalmbach committed:
Update jenkins file.

jbkalmbach committed:
Documentation update.

jbkalmbach committed:
Update tests and docs.

teweitsai commented on policy/phosimCmptSetting.yaml:
```diff
@@ -13,6 +13,10 @@ intraDirName: intra
 
+# Wavefront Sensor directory name. This is needed if the defocal image is
+# provided by the wavefront sensors.
+wfsDirName: wfs
```
If this is the directory name, where is the path (or root) in the file system? Or this is compared with the root of this repository?

jbkalmbach replied:
This is where PhoSim works and saves the images within each `iterX/img` directory.

teweitsai replied:
Please help to add the description to the comment. Thanks!

teweitsai commented on python/lsst/ts/phosim/CloseLoopTask.py:
```diff
@@ -715,6 +745,58 @@ def _calcWfErrFromImg(
             Observation ID used in PhoSim.
         butlerRootPath : str
             Path to the butler repository.
+        instName : `str`
```
instName : `str` --> instName : str

teweitsai commented on python/lsst/ts/phosim/CloseLoopTask.py:
```diff
@@ -715,6 +745,58 @@ def _calcWfErrFromImg(
+            Observation ID used in PhoSim.
+        butlerRootPath : str
+            Path to the butler repository.
+        instName : `str`
```
instName : `str` --> instName : str

teweitsai commented on python/lsst/ts/phosim/CloseLoopTask.py:
```diff
@@ -753,14 +835,98 @@ def _calcWfErrFromImg(
+
+        Parameters
+        ----------
+        obsId : `int`
+            Observation id.
+        butlerRootPath : `str`
+            Path to the butler gen3 repos.
+        instName : `str`
+            Instrument name.
+        numPro : int, optional
+            Number of processor to run DM pipeline. (the default is 1.)
+
+        Returns
+        -------
+        listOfWfErr : `list` of `SensorWavefrontError`
```
Please be consistent that you want to add `` to data type or not (see `numPro`).

teweitsai commented on python/lsst/ts/phosim/CloseLoopTask.py:
```diff
@@ -1013,7 +1194,15 @@ def runImg(
+            self.phosimCmpt.tele.setInstName(camType, defocalDist=0.0)
+            self.ofcCalc.ofc_data.normalized_image_quality_weight = np.append(
+                self.ofcCalc.ofc_data.normalized_image_quality_weight,
+                [0.25, 0.25, 0.25, 0.25],
```
Please leave a comment to explain this magic number: [0.25, 0.25, 0.25, 0.25]. In the future, we may want to put this into the configuration file or use a new algorithm to adjust the weighting in the run time.

teweitsai commented on python/lsst/ts/phosim/OpdMetrology.py:
```diff
@@ -195,6 +195,16 @@ def setDefaultLsstGQ(self):
+        """
+
+        # Set Equal Weights for WFS
+        self.setWeightingRatio([1.0, 1.0, 1.0, 1.0])
```
Please have a comment to explain: `[1.0, 1.0, 1.0, 1.0]`.

teweitsai commented on python/lsst/ts/phosim/PhosimCmpt.py:
```diff
@@ -715,6 +729,79 @@ def getPistonCamStarArgsAndFilesForPhoSim(
+
+        Parameters
+        ----------
+        obsId : int
+            Observation Id.
+        skySim : SkySim
+            Sky simulator
+        cmdFileName : str, optional
+            Physical command file name. (the default is "star.cmd".)
+        instFileName : str, optional
+            Star instance file name. (the default is "star.inst".)
+        logFileName : str, optional
+            Log file name. (the default is "starPhoSim.log".)
+        simSeed : int, optional
+            Random number seed. (the default is 1000)
+        cmdSettingFileName : str, optional
+            Physical command setting file name. (the default is
+            "starDefault.cmd".)
+        instSettingFileName : str, optional
+            Instance setting file name. (the default is "starSingleExp.inst".)
```
The order should be `obsId`, `skySim`, `simSeed`, `cmdSettingFileName`, and `instSettingFileName`. In addition, you should not have `cmdFileName`, `instFileName`, and `logFileName`.

teweitsai commented on python/lsst/ts/phosim/PhosimCmpt.py:
```diff
@@ -1198,6 +1288,45 @@ def getListOfFwhmSensorData(self, pssnFileName, sensorIdList):
+
+        Parameters
+        ----------
+        instName : `str`
```
instName : `str` --> instName : str

teweitsai commented on python/lsst/ts/phosim/PhosimCmpt.py:
```diff
@@ -1198,6 +1288,45 @@ def getListOfFwhmSensorData(self, pssnFileName, sensorIdList):
+        argstring += f" --inst {instName} "
+        if isEimg:
+            argstring += " --eimage"
+        argstring += " --focusz 0"
```
You may want to explain the value `0` here.

teweitsai commented on python/lsst/ts/phosim/telescope/TeleFacade.py:
```diff
@@ -324,14 +324,22 @@ def setInstName(self, camType, defocalDist=None):
 
-        if defocalDist <= 0:
-            raise ValueError("Defocal distance can not be <= 0.")
+        if (defocalDist == 0) & (camType != CamType.LsstCam):
```
`if (defocalDist == 0) & (camType != CamType.LsstCam):` --> `if (defocalDist == 0) and (camType != CamType.LsstCam):`

teweitsai reviewed (approved):
Looks good and thanks for this update! Please help to double check the closed-loop simulation of ComCam and LsstFam can still run well. Thanks!

jbkalmbach committed:
Update documentation after review by Te-Wei.
Fix bug preventing lsstfam and comcam from running properly after cwfs changes.

jbkalmbach commented:
Verified that this now runs with `comcam` and `lsstfam` mode as well.
